plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,553 @@
+						   error))
+			return FALSE;
+
+		if ((i + 1) % 4 == 0 || (i + 1) == chunks_payload->len) {
+			if (!fu_device_retry_full(FU_DEVICE(self),
+						  fu_algoltek_device_status_check,
+						  10,
+						  0,
+						  NULL,
+						  error))
+				return FALSE;
+		}
+		fu_progress_step_done(progress);
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usb_device_setup(FuDevice *device, GError **error)
+{
+	FuAlgoltekUsbDevice *self = FU_ALGOLTEK_USB_DEVICE(device);
+	g_autofree gchar *version_str = NULL;
+	g_autoptr(GByteArray) version_data = NULL;
+
+	/* UsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_algoltek_usb_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	version_data = algoltek_device_rdv(self, error);
+	version_str = g_strdup_printf("%s", version_data->data);
+
+	fu_device_set_version(device, version_str);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usb_device_write_firmware(FuDevice *device,
+				      FuFirmware *firmware,
+				      FuProgress *progress,
+				      FwupdInstallFlags flags,
+				      GError **error)
+{
+	FuAlgoltekUsbDevice *self = FU_ALGOLTEK_USB_DEVICE(device);
+	g_autoptr(GBytes) blob_isp = NULL;
+	g_autoptr(GBytes) blob_payload = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 18, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 2, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 80, NULL);
+
+	if (!algoltek_device_en(self, error)) {
+		g_prefix_error(error, "system activation failure: ");
+		return FALSE;
+	}
+
+	if (!algoltek_device_rst(self, 0x200, error)) {
+		g_prefix_error(error, "system reboot failure: ");
+		return FALSE;
+	}
+
+	fu_device_sleep(FU_DEVICE(self), 900);
+
+	if (!algoltek_device_wrr(self, 0x80AD, 0, error)) {
+		g_prefix_error(error, "data write failure: ");
+		return FALSE;
+	}
+
+	if (!algoltek_device_wrr(self, 0x80C0, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80C9, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80D1, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80D9, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80E1, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80E9, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_rst(self, 0, error)) {
+		g_prefix_error(error, "system reboot failure: ");
+		return FALSE;
+	}
+
+	fu_device_sleep(FU_DEVICE(self), 500);
+
+	/* get ISP image */
+	blob_isp = fu_firmware_get_image_by_id_bytes(firmware, "ISP", error);
+	if (blob_isp == NULL)
+		return FALSE;
+	if (!algoltek_device_ISP(self,
+				 blob_isp,
+				 AG_ISP_ADDR,
+				 fu_progress_get_child(progress),
+				 error)) {
+		g_prefix_error(error, "isp failure: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	if (!algoltek_device_bot(self, AG_ISP_ADDR, error)) {
+		g_prefix_error(error, "system boot failure: ");
```

Comment:
This and other prefixes for the helpers I think make more sense in the helpers themselves rather than the caller.

Reply:
Would you mind sharing more informastion with us about the helper? Actually we have no idea about it from other vendor's plugin.

Reply:
Yeah I think this makes more sense.

Reply:
Already updated!